### PR TITLE
[Improvement]: Generate return types for setters in data objects

### LINKED ISSUE #12185: [Improvement]: Generate return types for setters in data objects
### Improvement description

We noticed, that the return type in the generated setter methods of the data objects is missing. Is this on purpose?
Otherwise, I would suggest adding that, if it's possible and does not create any problems with the class overrides.  
(e.g.: `public function setSeries(?string $series): self {...}`)

Example of the current state (from the Pimcore demo):

```php
/**
* Set series - Series
* @param string|null $series
* @return \Pimcore\Model\DataObject\Car
*/
public function setSeries(?string $series)
{
	$this->series = $series;

	return $this;
}
```

### PR DESCRIPTION
**Pull request by mcop1**
Resolves #12185

### CODE REVIEW

File: models/DataObject/ClassDefinition/Data.php [hunk 1]
```diff
@@ -1075,7 +1075,7 @@ public function getSetterCodeLocalizedfields($class)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
         $code .= '*/' . "\n";
-        $code .= 'public function set' . ucfirst($key) . ' (' . $typeDeclaration . '$' . $key . ', $language = null)' . "\n";
+        $code .= 'public function set' . ucfirst($key) . ' (' . $typeDeclaration . '$' . $key . ', $language = null): \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= 'public function set' . ucfirst($key) . ' (' . $typeDeclaration . '$' . $key . ', $language = null): static' . "\n";
```

File: models/DataObject/ClassDefinition/Data.php [hunk 2]
```diff
@@ -686,7 +686,7 @@ public function getSetterCode($class)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
         $code .= '*/' . "\n";
-        $code .= 'public function set' . ucfirst($key) . '(' . $typeDeclaration . '$' . $key . ')' . "\n";
+        $code .= 'public function set' . ucfirst($key) . '(' . $typeDeclaration . '$' . $key . '): \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= 'public function set' . ucfirst($key) . '(' . $typeDeclaration . '$' . $key . '): static' . "\n";
```

File: lib/DataObject/ClassBuilder/ObjectBrickContainerClassBuilder.php [hunk 3]
```diff
@@ -80,7 +80,7 @@ public function buildContainerClass(Definition $definition, ClassDefinition $cla
             $cd .= '* @param \\Pimcore\\Model\\DataObject\\Objectbrick\\Data\\' . ucfirst($brickKey) . ' $' . $brickKey . "\n";
             $cd .= '* @return \\'.$namespace.'\\'.$className."\n";
             $cd .= '*/' . "\n";
-            $cd .= 'public function set' . ucfirst($brickKey) . '(' . '$' . $brickKey . ')' . "\n";
+            $cd .= 'public function set' . ucfirst($brickKey) . '(' . '$' . $brickKey . '): \\'.$namespace.'\\'.$className."\n";
```
Comment (blankse) [inline lib/DataObject/ClassBuilder/ObjectBrickContainerClassBuilder.php]:
```suggestion
            $cd .= 'public function set' . ucfirst($brickKey) . '(' . '$' . $brickKey . '): static' . "\n";
```

File: models/DataObject/ClassDefinition/Data/CalculatedValue.php [hunk 4]
```diff
@@ -445,7 +445,7 @@ public function getSetterCode($class)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\' . ucfirst($class->getName()) . "\n";
         $code .= '*/' . "\n";
-        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . ')' . "\n";
+        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . '): \\Pimcore\\Model\\DataObject\\' . ucfirst($class->getName()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/CalculatedValue.php]:
```suggestion
        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . '): static' . "\n";
```

File: models/DataObject/ClassDefinition/Data/CalculatedValue.php [hunk 5]
```diff
@@ -514,7 +514,7 @@ public function getSetterCodeLocalizedfields($class)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
         $code .= '*/' . "\n";
-        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . ', $language = null)' . "\n";
+        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . ', $language = null): \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/CalculatedValue.php]:
```suggestion
        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . ', $language = null): static' . "\n";
```

**Comment by blankse** [comment]
The `@return` PhpDoc should be changed to `@return $this`

File: models/DataObject/ClassDefinition/Data.php [hunk 6]
```diff
@@ -820,7 +820,7 @@ public function getSetterCodeObjectbrick($brickClass)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\Objectbrick\\Data\\' . ucfirst($brickClass->getKey()) . "\n";
         $code .= '*/' . "\n";
-        $code .= 'public function set' . ucfirst($key) . ' (' . $typeDeclaration . '$' . $key . ')' . "\n";
+        $code .= 'public function set' . ucfirst($key) . ' (' . $typeDeclaration . '$' . $key . '): \\Pimcore\\Model\\DataObject\\Objectbrick\\Data\\' . ucfirst($brickClass->getKey()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= 'public function set' . ucfirst($key) . ' (' . $typeDeclaration . '$' . $key . '): static' . "\n";
```

File: models/DataObject/ClassDefinition/Data.php [hunk 7]
```diff
@@ -820,7 +820,7 @@ public function getSetterCodeObjectbrick($brickClass)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\Objectbrick\\Data\\' . ucfirst($brickClass->getKey()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= '* @return $this' . "\n";
```

File: models/DataObject/ClassDefinition/Data.php [hunk 8]
```diff
@@ -948,7 +948,7 @@ public function getSetterCodeFieldcollection($fieldcollectionDefinition)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\Fieldcollection\\Data\\' . ucfirst($fieldcollectionDefinition->getKey()) . "\n";
         $code .= '*/' . "\n";
-        $code .= 'public function set' . ucfirst($key) . '(' . $typeDeclaration . '$' . $key . ')' . "\n";
+        $code .= 'public function set' . ucfirst($key) . '(' . $typeDeclaration . '$' . $key . '): \\Pimcore\\Model\\DataObject\\Fieldcollection\\Data\\' . ucfirst($fieldcollectionDefinition->getKey()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= 'public function set' . ucfirst($key) . '(' . $typeDeclaration . '$' . $key . '): static' . "\n";
```

File: models/DataObject/ClassDefinition/Data.php [hunk 9]
```diff
@@ -948,7 +948,7 @@ public function getSetterCodeFieldcollection($fieldcollectionDefinition)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\Fieldcollection\\Data\\' . ucfirst($fieldcollectionDefinition->getKey()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= '* @return $this' . "\n";
```

File: models/DataObject/ClassDefinition/Data.php [hunk 10]
```diff
@@ -1075,7 +1075,7 @@ public function getSetterCodeLocalizedfields($class)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= '* @return $this' . "\n";
```

File: models/DataObject/ClassDefinition/Data.php [hunk 11]
```diff
@@ -1155,7 +1155,7 @@ public function getFilterCode()
         $code .= '* @return static'."\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= '* @return $this'."\n";
```

File: models/DataObject/ClassDefinition/Data/CalculatedValue.php [hunk 12]
```diff
@@ -445,7 +445,7 @@ public function getSetterCode($class)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\' . ucfirst($class->getName()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/CalculatedValue.php]:
```suggestion
        $code .= '* @return $this' . "\n";
```

File: models/DataObject/ClassDefinition/Data/CalculatedValue.php [hunk 13]
```diff
@@ -467,7 +467,7 @@ public function getSetterCodeObjectbrick($brickClass)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\Objectbrick\\Data\\' . ucfirst($brickClass->getKey()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/CalculatedValue.php]:
```suggestion
        $code .= '* @return $this' . "\n";
```

File: models/DataObject/ClassDefinition/Data/CalculatedValue.php [hunk 14]
```diff
@@ -467,7 +467,7 @@ public function getSetterCodeObjectbrick($brickClass)
...
         $code .= '*/' . "\n";
-        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . ')' . "\n";
+        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . '): \\Pimcore\\Model\\DataObject\\Objectbrick\\Data\\' . ucfirst($brickClass->getKey()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/CalculatedValue.php]:
```suggestion
        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . '): static' . "\n";
```

File: models/DataObject/ClassDefinition/Data/CalculatedValue.php [hunk 15]
```diff
@@ -487,7 +487,7 @@ public function getSetterCodeFieldcollection($fieldcollectionDefinition)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\Fieldcollection\\Data\\' . ucfirst($fieldcollectionDefinition->getKey()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/CalculatedValue.php]:
```suggestion
        $code .= '* @return $this' . "\n";
```

File: models/DataObject/ClassDefinition/Data/CalculatedValue.php [hunk 16]
```diff
@@ -487,7 +487,7 @@ public function getSetterCodeFieldcollection($fieldcollectionDefinition)
...
         $code .= '*/' . "\n";
-        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . ')' . "\n";
+        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . '): \\Pimcore\\Model\\DataObject\\Fieldcollection\\Data\\' . ucfirst($fieldcollectionDefinition->getKey()) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/CalculatedValue.php]:
```suggestion
        $code .= 'public function set' . ucfirst($key) . '(' . '$' . $key . '): static' . "\n";
```

File: models/DataObject/ClassDefinition/Data/CalculatedValue.php [hunk 17]
```diff
@@ -514,7 +514,7 @@ public function getSetterCodeLocalizedfields($class)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data/CalculatedValue.php]:
```suggestion
        $code .= '* @return $this' . "\n";
```

File: models/DataObject/ClassDefinition/Data.php [hunk 18]
```diff
@@ -686,7 +686,7 @@ public function getSetterCode($class)
         $code .= '* @param ' . $this->getPhpdocInputType() . ' $' . $key . "\n";
         $code .= '* @return \\Pimcore\\Model\\DataObject\\' . ucfirst($classname) . "\n";
```
Comment (blankse) [inline models/DataObject/ClassDefinition/Data.php]:
```suggestion
        $code .= '* @return $this' . "\n";
```

File: lib/DataObject/ClassBuilder/ObjectBrickContainerClassBuilder.php [hunk 19]
```diff
@@ -80,7 +80,7 @@ public function buildContainerClass(Definition $definition, ClassDefinition $cla
             $cd .= '* @param \\Pimcore\\Model\\DataObject\\Objectbrick\\Data\\' . ucfirst($brickKey) . ' $' . $brickKey . "\n";
             $cd .= '* @return \\'.$namespace.'\\'.$className."\n";
```
Comment (blankse) [inline lib/DataObject/ClassBuilder/ObjectBrickContainerClassBuilder.php]:
```suggestion
            $cd .= '* @return $this' . "\n";
```

**Comment by mcop1** [comment]
> The `@return` PhpDoc should be changed to `@return $this`

@blankse you are right, thank you for the suggestion(s)!

**Pull request merged**